### DISCUSSION TITLE
sqlpage.exec : capturing stderr and result code

### DISCUSSION THREAD

**Discussion by guspower** [sqlpage/SQLPage#291] [Ideas]
Hi,

After reading through [sqlpage.exec](https://sql.ophir.dev/functions.sql?function=exec#function) and trying it out, I'm wondering if it's possible to somehow capture the `stderr` and `exit code`. The scenarios I specifically want to deal with are where the executable fails. I had considered writing a wrapper to combine stderr and stdout (e.g. `2>&1`) but I was interested to know if anyone else has a solution for this.

**Comment by lovasoa** [comment]
Hello ! 

`sqlpage.exec` currently does not allow that. Here is a small python wrapper that does what you need:

### `execute_to_json.py`

```py
import subprocess, sys, json
res = subprocess.run(sys.argv[1:], capture_output=True, text=True)
print(json.dumps({
  "stdout": res.stdout,
  "stderr": res.stderr,
  "exit_code": res.returncode
}))
``` 

### `index.sql`

```sql
set execution_results = sqlpage.exec('python3', '/usr/local/bin/execute_to_json.py', 'echo', 'hello world');

select 'text' as component,
  CONCAT(
        'Standard output: ', $execution_results->>'stdout',
        'Standard error: ', $execution_results->>'stderr',
        'Exit code: ', $execution_results->>'exit_code'
  ) as contents;
```
